### PR TITLE
[`absa`] Add SetFitABSA notebook on FiQA

### PR DESCRIPTION
**Pull request by tomaarsen**
Hello!

## Pull Request overview
* Add SetFitABSA notebook on FiQA

## Details
The initial work was carried out by @rlaperdo @MosheWasserb @orenpereg. I made a few changes here and there, e.g. I added an inference section & clarified some things.

- Tom Aarsen

### CODE REVIEW

File: notebooks/setfit-absa-fiqa.ipynb [hunk 1]
```diff
@@ -0,0 +1,7780 @@
+{
```
Comment (rlaperdo) [inline notebooks/setfit-absa-fiqa.ipynb:1]: <p>In their paper [BloombergGPT: A Large Language Model for Finance](<a href="https://arxiv.org/pdf/2303.17564.pdf)" target="_blank">https://arxiv.org/pdf/2303.17564.pdf)</a>, Bloomberg presents the development of <strong>**BloombergGPT**</strong>, LLM which is specialized for the financial domain.&lt;br&gt;</p><p>The model is a <strong>**50B**</strong> parameters language model, trained on a wide range of financial data:</p><p> </p><p>* &nbsp; <strong>**363 billion**</strong> tokens dataset based on Bloomberg's extensive data sources</p><p>* &nbsp; <strong>**345 billion**</strong> tokens from general purpose datasets</p><p> </p><p>The model was evaluated on various financial NLP tasks, including Aspect-Based Sentiment Analysis (<strong>**ABSA**</strong>). Bloomberg evaluated ABSA in an in-context manner using the [FiQA_SA](<a href="https://huggingface.co/datasets/AdaptLLM/finance-tasks/viewer/FiQA_SA)" target="_blank">https://huggingface.co/datasets/AdaptLLM/finance-tasks/viewer/FiQA_SA)</a>. This dataset contains a test set of 235 finance related sentences, each is prefixed by 5 tagged sentences, which overall comprise an input prompt to the model. Every tagged sentence in the prompt contains a known aspect and a question whether its corresponding polarity is Positive, Negative or Neutral, and the answer (the correct polarity towards the mentioned aspect) is written at the end. BloombergGPT is expected to predict the polarity of a given aspect in the test sentence based on the 5 tagged example sentences. This is an <strong>**SB2 ABSA task**</strong>, which means that the aspect is already given and the model just needs to predict the corresponding polarity.</p><p> </p><p>The evaluation score achieved by this model is <strong>**weighted F1-score=75.07**</strong>, as reported in table 8 in the paper.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/setfit/pull/471/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (rlaperdo) [inline notebooks/setfit-absa-fiqa.ipynb:1]: <p>[SetFitABSA](<a href="https://huggingface.co/blog/setfit-absa)" target="_blank">https://huggingface.co/blog/setfit-absa)</a></p><p>is a framework for <strong>**few-shot**</strong> training of domain-specific ABSA models. It provides an efficient and accurate technique to detect sentiment towards specific aspects within the text. The following graph shows evaluation results of a SetFitABSA model that was trained with just few k finance related sentences and tested on the same 235 FiQA_SA test sentences that were used used to evaluate BloombergGPT. The size of the model is based on its underlying sentence transformer and is usually considerably smaller compared to SOTA LLMs.</p><p>Specifically, in this work, we used the <strong>**paraphrase-mpnet-base-v2**</strong> sentence transformer which contains only <strong>**110M**</strong> params compared to <strong>**50B**</strong> of BloombergGPT.&lt;br&gt;</p><p> </p><p>We separated from each prompt in [FiQA_SA](<a href="https://huggingface.co/datasets/AdaptLLM/finance-tasks/viewer/FiQA_SA)" target="_blank">https://huggingface.co/datasets/AdaptLLM/finance-tasks/viewer/FiQA_SA)</a> the tagged and non-tagged sentences, to create [train and test sets](<a href="https://huggingface.co/datasets/ronenlap/SetFitAbsa_FiQA)" target="_blank">https://huggingface.co/datasets/ronenlap/SetFitAbsa_FiQA)</a> correspondingly. The train set contains 646 unique sentences overall.&lt;br&gt;This graph summarizes the evaluation results. It presents the weighted F1-score achieved by SetFitABSA as a function of k, the number of sampled training sentences. The result for each k is averaged over 5 seeds. We can see that with only 50 tranining sentences SetFitABSA exceeds the BloombergGPT score. BTW, if we use the full training set for training the SetFitABSA model we get a weighted F1-score of over 86.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/setfit/pull/471/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (rlaperdo) [inline notebooks/setfit-absa-fiqa.ipynb:1]: <p># **Demo Code**</p><p>The following demonstrates the training of the SetFitABSA model and its evaluation over the <strong>**FiQA_SA**</strong> financial dataset.</p><p> </p><p>Specifically:</p><p> </p><p>* &nbsp; We sample <strong>**k=24**</strong> sentences from the training set using an arbitrary seed</p><p>* &nbsp; We trained the SetFitABSA model using only these <strong>**k**</strong> samples</p><p>* &nbsp; We evaluate the trained model over the entire test set which is exactly the same set that was used to evalute BloombergGPT model. The evaluated task is <strong>**SB2**</strong> - aspects are given and the model needs to predict their corresponding polarity</p><p> </p><p>With just 24 training sentences we get a weighted F1-score which is on par or better than the 75.07 score achieved by BloombergGPT (try it yourself!)</p><p>&lt;br&gt;</p><p>&lt;br&gt;</p><p>&lt;br&gt;</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/huggingface/setfit/pull/471/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>
